### PR TITLE
Fixes #28217 - drop turbolinks

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Drops turbolinks - takes over from #7158 (thanks @amirfefer :+1:)

### CODE REVIEW

File: docker-compose.yml
Comment (ekohl) [inline docker-compose.yml]: This looks incorrect

File: webpack/assets/javascripts/dashboard/index.js
Comment (ekohl) [inline webpack/assets/javascripts/dashboard/index.js]: This will be painful I suspect. We default it to off nowadays so I suspect many users won't notice this. Probably deserves some attention.

File: webpack/assets/javascripts/foreman_tools.js
Comment (ekohl) [inline webpack/assets/javascripts/foreman_tools.js]: Since 184f8d5ce23cfe208a696535bcea346074d8600a the entire `updateTable` function is unused in Foreman itself. Don't know about plugins but I'd appreciate it if someone would at least deprecate it but possibly just remove it.

File: webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js
Comment (ekohl) [inline webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js]: Not a fan of commented code. I'd prefer to drop it.

**Review by ekohl — COMMENTED** [review]
Looks like this will affect the dashboard when autorefresh is enabled and searching (via `changeQuery`).

File: webpack/assets/javascripts/dashboard/index.js
Comment (ezr-ondrej) [reply webpack/assets/javascripts/dashboard/index.js]: I would guess it was painfull already. Could we just solve it without refreshing the page?

File: docker-compose.yml
Comment (ezr-ondrej) [reply docker-compose.yml]: :man_facepalming: Thanks, it definitely doesn't belong here.

File: webpack/assets/javascripts/foreman_tools.js
Comment (ezr-ondrej) [reply webpack/assets/javascripts/foreman_tools.js]: I've deprecated it, just in case.

File: webpack/assets/javascripts/foreman_tools.test.js
Comment (sharvit) [inline webpack/assets/javascripts/foreman_tools.test.js]: Could use the `mockWindowLocation` here?

File: webpack/assets/javascripts/foreman_tools.js
Comment (sharvit) [inline webpack/assets/javascripts/foreman_tools.js]: `window.URL_PREFIX` instead of the comment would work?

File: webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js:42]: Shouldn't we expect something after pressing ENTER?

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @ezr-ondrej

File: webpack/assets/javascripts/dashboard/index.js
Comment (ekohl) [reply webpack/assets/javascripts/dashboard/index.js]: I think the dashboard is due for a proper rewrite to React and only refresh relevant content. However, that needs some thought. Most of the graphs there aren't that interesting to refresh regularly because in many use cases they don't change over the course of days. AFAIK we default to not enabling auto refresh now so it doesn't block this effort. Just good to be aware of the bigger picture.

File: webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js
Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js:42]: I don't know what to expect though :thinking: the url should stay the same :slightly_frowning_face:

**Comment by ohadlevy** [comment]
does it make sense to have a tubolink.visit function that is deprecated (in case some other plugin is using it?)

**Comment by ezr-ondrej** [comment]
> does it make sense to have a tubolink.visit function that is deprecated (in case some other plugin is using it?)

Good point, it surely does make sense.

File: webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js:42]: Would it make sense spying on the `global.window.location.href` setter?
https://jestjs.io/docs/en/jest-object#jestspyonobject-methodname-accesstype

Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js:42]: I needed a bit of codewrangling. Not sure if this is the best way, but it works :thinking: :laughing:

File: app/assets/javascripts/application.js
Comment (tbrisker) [inline app/assets/javascripts/application.js]: This (and the above) is a turbolinks event - we should also remove anything that hooks into these events. The full list is at: https://github.com/turbolinks/turbolinks-classic#events

File: webpack/assets/javascripts/foreman_tools.js
Comment (tbrisker) [inline webpack/assets/javascripts/foreman_tools.js]:
```suggestion
  deprecate('updateTable', 'react Table component', '2.1');
```

File: webpack/assets/javascripts/foreman_globals.js
Comment (tbrisker) [inline webpack/assets/javascripts/foreman_globals.js]:
```suggestion
      '2.1'
```

**Review by tbrisker — CHANGES_REQUESTED** [review]
a few inline comments, otherwise good effort!

File: app/assets/javascripts/application.js
Comment (ezr-ondrej) [reply app/assets/javascripts/application.js]: I've done it for core, as there was no funcionality hook on it (only cleanups), it was easy.

File: webpack/assets/javascripts/foreman_globals.js
Comment (sharvit) [inline webpack/assets/javascripts/foreman_globals.js]: Maybe it would make more sense to put it in `foreman_navigation`?

File: webpack/assets/javascripts/react_app/common/MountingService.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/MountingService.js]: Should we keep managing this list now?
The whole point of this list was to be able to `unmount` all the `mountedNodes` before turbolinks mount them again.

File: webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js:42]: LGTM :+1:

File: webpack/assets/javascripts/foreman_navigation.js
Comment (sharvit) [inline webpack/assets/javascripts/foreman_navigation.js:8]: Sorry for noticing it just now but moving it to the `foreman_navigation.js` make me wonder.
Should we add something like?
```js
export const visit = url => {
 window.location.href = url;
};

// from legacy js
window.tfm.nav.visit('some-url');

// from webpack
import { visit } from '../foreman_navigation';

visit('some-url');
```

It might help in a future refactoring.

Comment (ezr-ondrej) [reply webpack/assets/javascripts/foreman_navigation.js:8]: Oh, right, thats nice! I do not really like it comming after I did it everywhere, but hey good idea, let's do it!

Comment (ezr-ondrej) [reply webpack/assets/javascripts/foreman_navigation.js:8]: Done that, it wasn't that bad :) and looks much better :wink:

File: webpack/assets/javascripts/dashboard/index.js
Comment (ekohl) [inline webpack/assets/javascripts/dashboard/index.js]: I think you meant this:
```suggestion
import { reloadPage } from '../foreman_navigation';
```

**Comment by sharvit** [comment]
Are the failing tests relevant?

File: bundler.d/assets.rb
Comment (ohadlevy) [inline bundler.d/assets.rb]: would be really nice to get rid of the x-editable-rails (yes i know, not in this pr :))

**Comment by ezr-ondrej** [comment]
> Are the failing tests relevant?

I believe not. Travis says just `eslint` missing. I don't believe that's relevant. And I think in jenkins just failed to start the server for integration-tests, but that could be related. let's see.

**Comment by ezr-ondrej** [comment]
[test foreman]

File: bundler.d/assets.rb
Comment (ezr-ondrej) [reply bundler.d/assets.rb]: Yeah, that would be nice, but do we have a replacement?

**Comment by ohadlevy** [comment]
I believe its a pretty simple react component? doesn't patternfly has in
place editing?

**Comment by ezr-ondrej** [comment]
> I believe its a pretty simple react component? doesn't patternfly has in place editing?

I can't find it in their storybook, but they have it in the patternfly itself

**Comment by sharvit** [comment]
> I can't find it in their storybook, but they have it in the patternfly itself

PF3 doesn't have it, PF4 have it in tables https://patternfly-react.surge.sh/patternfly-4/documentation/react/inline-table/inlineedit

**Comment by tbrisker** [comment]
pushed rebase to make sure tests are green

File: webpack/assets/javascripts/react_app/routes/index.js
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/routes/index.js:26]: This is now causing an endless loop, since `currentPath` is reset to null on page load

Comment (tbrisker) [reply webpack/assets/javascripts/react_app/routes/index.js:26]: pushed a commit that should fix this

**Comment by tbrisker** [comment]
travis failure is on node10 eslint and should be fixed by https://github.com/theforeman/foreman/pull/7225

**Review by tbrisker — DISMISSED** [review]
Fine by me but would be good to get another ack at least on the additional commit i pushed before merging.

File: bundler.d/assets.rb
Comment (tbrisker) [inline bundler.d/assets.rb]: We need to package this gem as well until we can get rid of it

Comment (mmoll) [reply bundler.d/assets.rb]: At the moment, 4.2.2 is packaged for RPM. Can we use this for now and update after https://github.com/theforeman/foreman-packaging/pull/4517 got in?

Comment (tbrisker) [reply bundler.d/assets.rb]: we'll still need 5.0 for rails 6 compat

Comment (mmoll) [reply bundler.d/assets.rb]: That why I said "for now" ;)

Comment (tbrisker) [reply bundler.d/assets.rb]: pushed another commit with both versions dependent on rails version

File: webpack/assets/javascripts/react_app/common/testHelpers.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/testHelpers.js:30]: Does it still need after we added `visit` to `foreman_navigation`?
looks like the `visit` method get mocked now instead and it looks better.

File: webpack/assets/javascripts/react_app/common/urlHelpers.test.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/urlHelpers.test.js:43]: Does it still need it here? Looks like the `visit` is the mock now

File: webpack/assets/javascripts/react_app/components/Layout/Layout.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: The `preventHref` was here so we can use turbolinks when needed, looks like it can be removed from more places now.

**Review by sharvit — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/common/urlHelpers.test.js
Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/common/urlHelpers.test.js:43]: Well the urlHelpers work with the `window.location.href` directly, so it is necessary here.

File: webpack/assets/javascripts/react_app/components/Layout/Layout.js
Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: Could you point me to an usage of this method? I can't see where we use it, or you mean the `preventHref` is used on different places as well? If so, could you point me to those? :) I found it only in `LayoutSelectors.js` something I've missed?

**Comment by ezr-ondrej** [comment]
I've rebased it again answered @sharvit's questions. Some other ideas?

Thanks @tbrisker for shepherding this PR :heart:

File: webpack/assets/javascripts/react_app/components/Layout/Layout.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: 1. The `LayoutSelectors.js` uses `item.turbolinks` received from the server:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/LayoutSelectors.js#L33
Not sure if this PR should remove it, if so, there are also the `data-turbolinks-track` we use in many places.
There is a menu-items registry where we define menu items to use turbolinks :-1: 
https://github.com/theforeman/foreman/blob/develop/app/registries/menu/item.rb#L21
Looks like `turbolinks` is default true, we just use `turbolinks` for everything.

2. Looks like the `preventHref` still needed so it will use the `onNavigate` for the navigation.
See: https://github.com/patternfly/patternfly-react/blob/dbff0f181a2bd6fd763fb6769e8f9ef49281178d/packages/patternfly-3/patternfly-react/src/components/VerticalNav/VerticalNavItemHelper.js

Comment (tbrisker) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: @ezr-ondrej looks like this comment is still not addressed? or am i missing something?

Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: oh, I've missed this one

Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: @sharvit we don't need to use the `preventHref` right?

File: app/registries/menu/item.rb
Comment (tbrisker) [inline app/registries/menu/item.rb]: do we even want to keep this option? doesn't seem like we should given it's not used. iirc this was originally introduced to allow skipping turbolinks on pages we know will cause a full page reload when loaded (due to js etc) but i'm not sure if it has ever been actually used.

File: webpack/assets/javascripts/react_app/components/Layout/Layout.js
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: do we actually need to use `onNavigate` that basically does the same as just letting the href do it's thing?

File: webpack/assets/javascripts/react_app/components/Layout/components/__snapshots__/NavDropdown.test.js.snap
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/Layout/components/__snapshots__/NavDropdown.test.js.snap]: is this intentional?

File: webpack/assets/javascripts/react_app/components/Layout/Layout.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: Shouldn't we set it to true?
If not the browser will handle it natively? see this code from inside pf:
```js
if (preventHref && !!onClick) {
  event.preventDefault();
}
```
https://github.com/patternfly/patternfly-react/blob/dbff0f181a2bd6fd763fb6769e8f9ef49281178d/packages/patternfly-3/patternfly-react/src/components/VerticalNav/VerticalNavItemHelper.js#L88-L90

File: webpack/assets/javascripts/react_app/components/Layout/Layout.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: Isn't it here so it will use the react-router `history` object?

Comment (tbrisker) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: good point, i was confusing it with the `window.history`

**Comment by mmoll** [comment]
foreman test failures related :broken_heart:

File: webpack/assets/javascripts/react_app/components/Layout/components/__snapshots__/NavDropdown.test.js.snap
Comment (tbrisker) [reply webpack/assets/javascripts/react_app/components/Layout/components/__snapshots__/NavDropdown.test.js.snap]: guess not since the tests are failing on this

File: webpack/assets/javascripts/react_app/components/Layout/Layout.js
Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: I thought we want that for pages not handled by react router, right? :thinking:

**Comment by ezr-ondrej** [comment]
[test katello]

File: webpack/assets/javascripts/react_app/components/Layout/components/__snapshots__/NavDropdown.test.js.snap
Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/components/Layout/components/__snapshots__/NavDropdown.test.js.snap]: no, not really ... just blind `-u` :(

File: app/registries/menu/item.rb
Comment (ezr-ondrej) [reply app/registries/menu/item.rb]: dropped the option. You're right, we do not want to keep it.

**Comment by ares** [comment]
tests are green now

**Review by tbrisker — CHANGES_REQUESTED** [review]
Tests, this works fine except on the audits page - when changing page in the pagination there, an endless loop is triggered

**Comment by tbrisker** [comment]
disregard, looks like the issue is actually in develop and I didn't switch branches before testing

**Review by tbrisker — DISMISSED** [review]
Thanks @ezr-ondrej ! this works as expected, navigation between pages works correctly. Please drop turbolinks gem from packaging.

**Comment by ekohl** [comment]
I have a script to update dependencies that I should really upstream somewhere: https://github.com/ekohl/theforeman-plumbing/blob/master/fm-packaging-update-deps. I'd recommend to merge the PR and then call the script.

**Comment by ezr-ondrej** [comment]
> I have a script to update dependencies that I should really upstream somewhere: https://github.com/ekohl/theforeman-plumbing/blob/master/fm-packaging-update-deps. I'd recommend to merge the PR and then call the script.

I don't see any dropping of deps in the script. Maybe I just don't see it, but let's drop it manually. It is just deleting files, right?

**Comment by ekohl** [comment]
> I don't see any dropping of deps in the script. Maybe I just don't see it, but let's drop it manually. It is just deleting files, right?

First you should update foreman.spec to drop the dependency on those packages. Since the script takes the dependencies in this repository and converts them, it should automatically drop them if they're dropped in the core repository. This PR updates `Gemfile` and `bundler.d/assets.rb` and the script will pick that up.

File: webpack/assets/javascripts/react_app/components/Layout/Layout.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: **I'm confused already but I think we should set it to `true`.**

We need the `preventDefault` when clicking a link so instead of using the native browser `href` mechanism, it would use the react-router.

If you would set it to `false`, wouldn't it use react-router while also using the native browser link?

Comment (tbrisker) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: so what you are suggesting is we don't actually need to pass it in to the `onNavigate` function (since it isn't using it anyways) but rather explicitly add another prop below that does
```jsx
preventDefault=true
```
instead of relying on `{...this.props}` to include it and pass it on?

Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/Layout/Layout.js:88]: Yes!

**Comment by ezr-ondrej** [comment]
Packaging dropping the dependency on turbolinks: theforeman/foreman-packaging#4576

**Review by tbrisker — COMMENTED** [review]
Tested, works well and looks like the preventHref is indeed preventing browser routing between react pages. 
Thanks @ezr-ondrej and all reviewers! Will merge once tests are green.

**Review by tbrisker — DISMISSED** [review]

**Comment by ezr-ondrej** [comment]
[test foreman][test katello] jenkins failure, reseting the tests

**Comment by tbrisker** [comment]
[test katello]

**Review by sharvit — APPROVED** [review]
Thanks @ezr-ondrej, looks good and work as expected :+1:

**Comment by tbrisker** [comment]
SQLite tests failed to execute for some reason, but this doesn't actually touch anything db-related so i'll merge.

**Pull request merged**
